File: packages/google_mlkit_barcode_scanning/ios/Classes/GoogleMlKitBarcodeScanningPlugin.m
#import <Foundation/Foundation.h>
#import "GoogleMlKitBarcodeScanningPlugin.h"
#import <MLKitBarcodeScanning/MLKitBarcodeScanning.h>
#import <google_mlkit_commons/GoogleMlKitCommonsPlugin.h>

#define channelName @"google_mlkit_barcode_scanning"
#define startBarcodeScanner @"vision#startBarcodeScanner"
#define closeBarcodeScanner @"vision#closeBarcodeScanner"

@implementation GoogleMlKitBarcodeScanningPlugin {
    NSMutableDictionary *instances;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:channelName
                                     binaryMessenger:[registrar messenger]];
    GoogleMlKitBarcodeScanningPlugin* instance = [[GoogleMlKitBarcodeScanningPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)init {
    self = [super init];
    if (self)
        instances = [NSMutableDictionary dictionary];
    return  self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:startBarcodeScanner]) {
        [self handleDetection:call result:result];
    } else if ([call.method isEqualToString:closeBarcodeScanner]) {
        NSString *uid = call.arguments[@"id"];
        [instances removeObjectForKey:uid];
        result(NULL);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (MLKBarcodeScanner*)initialize:(FlutterMethodCall *)call {
    NSArray *array = call.arguments[@"formats"];
    NSInteger formats = 0;
    for (NSNumber *num in array) {
        formats += [num intValue];
    }
    MLKBarcodeScannerOptions *options = [[MLKBarcodeScannerOptions alloc] initWithFormats: formats];
    return [MLKBarcodeScanner barcodeScannerWithOptions:options];
}

- (void)handleDetection:(FlutterMethodCall *)call result:(FlutterResult)result {
    MLKVisionImage *image = [MLKVisionImage visionImageFromData:call.arguments[@"imageData"]];
    
    NSString *uid = call.arguments[@"id"];
    MLKBarcodeScanner *barcodeScanner = [instances objectForKey:uid];
    if (barcodeScanner == NULL) {
        barcodeScanner = [self initialize:call];
        instances[uid] = barcodeScanner;
    }
    
    [barcodeScanner processImage:image
                      completion:^(NSArray<MLKBarcode *> *barcodes,
                                   NSError *error) {
        if (error) {
            result(getFlutterError(error));
            return;
        } else if (!barcodes) {
            result(@[]);
            return;
        }
        
        NSMutableArray *array = [NSMutableArray array];
        for (MLKBarcode *barcode in barcodes) {
            [array addObject:[self barcodeToDictionary:barcode]];
        }
        result(array);
    }];
}

- (NSDictionary *)barcodeToDictionary:(MLKBarcode *)barcode {
    NSMutableDictionary *dictionary = [NSMutableDictionary new];
    [dictionary addEntriesFromDictionary:@{
        @"type" : @(barcode.valueType) ?: [NSNull null],
        @"format" : @(barcode.format) ?: [NSNull null],
        @"rawValue" : barcode.rawValue ?: [NSNull null],
        @"rawBytes" : barcode.rawData ?: [NSNull null],
        @"displayValue" : barcode.displayValue ?: [NSNull null],
        @"rect" : @{
            @"left" : @(barcode.frame.origin.x),
            @"top" : @(barcode.frame.origin.y),
            @"right" : @(barcode.frame.origin.x + barcode.frame.size.width),
            @"bottom" : @(barcode.frame.origin.y + barcode.frame.size.height)
        }
    }];
    
    NSMutableArray *cornerPoints = [NSMutableArray array];
    for (NSValue * point in barcode.cornerPoints) {
        CGPoint newPoint = [point CGPointValue];
        [cornerPoints addObject: @{
            @"x": @(newPoint.x),
            @"y": @(newPoint.y),
        }];
    }
    dictionary[@"points"] = cornerPoints;
    
    switch (barcode.valueType) {
        case MLKBarcodeValueTypeUnknown:
        case MLKBarcodeValueTypeISBN:
        case MLKBarcodeValueTypeProduct:
        case MLKBarcodeValueTypeText:
            break;
        case MLKBarcodeValueTypeWiFi:
            [dictionary addEntriesFromDictionary:[self wifiToDictionary:barcode.wifi]];
            break;
        case MLKBarcodeValueTypeURL:
            [dictionary addEntriesFromDictionary:[self urlToDictionary:barcode.URL]];
            break;
        case MLKBarcodeValueTypeEmail:
            [dictionary addEntriesFromDictionary:[self emailToDictionary:barcode.email]];
            break;
        case MLKBarcodeValueTypePhone:
            [dictionary addEntriesFromDictionary:[self phoneToDictionary:barcode.phone]];
            break;
        case MLKBarcodeValueTypeSMS:
            [dictionary addEntriesFromDictionary:[self smsToDictionary:barcode.sms]];
            break;
        case MLKBarcodeValueTypeGeographicCoordinates:
            [dictionary addEntriesFromDictionary:[self geoPointToDictionary:barcode.geoPoint]];
            break;
        case MLKBarcodeValueTypeDriversLicense:
            [dictionary addEntriesFromDictionary:[self driverLicenseToDictionary:barcode.driverLicense]];
            break;
        case MLKBarcodeValueTypeContactInfo:
            [dictionary addEntriesFromDictionary:[self contactInfoToDictionary:barcode.contactInfo]];
            break;
        case MLKBarcodeValueTypeCalendarEvent:
            [dictionary addEntriesFromDictionary:[self calendarEventToDictionary:barcode.calendarEvent]];
            break;
    }
    
    return dictionary;
}

- (NSDictionary *)wifiToDictionary:(MLKBarcodeWiFi *)wifi {
    return @{
        @"ssid" : wifi.ssid ?: [NSNull null],
        @"password" : wifi.password ?: [NSNull null],
        @"encryption" : @(wifi.type)
    };
}

- (NSDictionary *)urlToDictionary:(MLKBarcodeURLBookmark *)url {
    return @{
        @"title" : url.title ?: [NSNull null],
        @"url" : url.url ?: [NSNull null]
    };
}

- (NSDictionary *)emailToDictionary:(MLKBarcodeEmail *)email {
    return @{
        @"address" : email.address ?: [NSNull null],
        @"body" : email.body ?: [NSNull null],
        @"subject" : email.subject ?: [NSNull null],
        @"emailType" : @(email.type)
    };
}

- (NSDictionary *)phoneToDictionary:(MLKBarcodePhone *)phone {
    return @{
        @"number" : phone.number ?: [NSNull null],
        @"phoneType" : @(phone.type)
    };
}

- (NSDictionary *)smsToDictionary:(MLKBarcodeSMS *)sms {
    return @{
        @"number" : sms.phoneNumber ?: [NSNull null],
        @"message" : sms.message ?: [NSNull null]
    };
}

- (NSDictionary *)geoPointToDictionary:(MLKBarcodeGeoPoint *)geo {
    return @{
        @"longitude" : @(geo.longitude),
        @"latitude" : @(geo.latitude)
    };
}

- (NSDictionary *)driverLicenseToDictionary:(MLKBarcodeDriverLicense *)license {
    return @{
        @"firstName" : license.firstName ?: [NSNull null],
        @"middleName" : license.middleName ?: [NSNull null],
        @"lastName" : license.lastName ?: [NSNull null],
        @"gender" : license.gender ?: [NSNull null],
        @"addressCity" : license.addressCity ?: [NSNull null],
        @"addressStreet" : license.addressStreet ?: [NSNull null],
        @"addressState" : license.addressState ?: [NSNull null],
        @"addressZip" : license.addressZip ?: [NSNull null],
        @"birthDate" : license.birthDate ?: [NSNull null],
        @"documentType" : license.documentType ?: [NSNull null],
        @"licenseNumber" : license.licenseNumber ?: [NSNull null],
        @"expiryDate" : license.expiryDate ?: [NSNull null],
        @"issueDate" : license.issuingDate ?: [NSNull null],
        @"country" : license.issuingCountry ?: [NSNull null]
    };
}

- (NSDictionary *)contactInfoToDictionary:(MLKBarcodeContactInfo *)contact {
    NSMutableArray<NSDictionary *> *addresses = [NSMutableArray array];
    [contact.addresses enumerateObjectsUsingBlock:^(MLKBarcodeAddress *_Nonnull address,
                                                    NSUInteger idx, BOOL *_Nonnull stop) {
        NSMutableArray<NSString *> *addressLines = [NSMutableArray array];
        [address.addressLines enumerateObjectsUsingBlock:^(NSString *_Nonnull addressLine,
                                                           NSUInteger idx, BOOL *_Nonnull stop) {
            [addressLines addObject:addressLine];
        }];
        [addresses addObject:@{@"addressLines" : addressLines, @"addressType" : @(address.type)}];
    }];
    
    NSMutableArray<NSDictionary *> *emails = [NSMutableArray array];
    [contact.emails enumerateObjectsUsingBlock:^(MLKBarcodeEmail *_Nonnull email,
                                                 NSUInteger idx, BOOL *_Nonnull stop) {
        [emails addObject:@{
            @"address" : email.address ?: [NSNull null],
            @"body" : email.body ?: [NSNull null],
            @"subject" : email.subject ?: [NSNull null],
            @"emailType" : @(email.type)
        }];
    }];
    
    NSMutableArray<NSDictionary *> *phones = [NSMutableArray array];
    [contact.phones enumerateObjectsUsingBlock:^(MLKBarcodePhone *_Nonnull phone,
                                                 NSUInteger idx, BOOL *_Nonnull stop) {
        [phones addObject:@{@"number" : phone.number ?: [NSNull null], @"phoneType" : @(phone.type)}];
    }];
    
    NSMutableArray<NSString *> *urls = [NSMutableArray array];
    [contact.urls
     enumerateObjectsUsingBlock:^(NSString *_Nonnull url, NSUInteger idx, BOOL *_Nonnull stop) {
        [urls addObject:url];
    }];
    return @{
        @"addresses" : addresses,
        @"emails" : emails,
        @"phones" : phones,
        @"urls" : urls,
        @"formattedName" : contact.name.formattedName ?: [NSNull null],
        @"firstName" : contact.name.first ?: [NSNull null],
        @"lastName" : contact.name.last ?: [NSNull null],
        @"middleName" : contact.name.middle ?: [NSNull null],
        @"prefix" : contact.name.prefix ?: [NSNull null],
        @"pronunciation" : contact.name.pronunciation ?: [NSNull null],
        @"suffix" : contact.name.suffix ?: [NSNull null],
        @"jobTitle" : contact.jobTitle ?: [NSNull null],
        @"organization" : contact.organization ?: [NSNull null]
    };
}

- (NSDictionary *)calendarEventToDictionary:(MLKBarcodeCalendarEvent *)calendar {
    return @{
        @"description" : calendar.eventDescription ?: [NSNull null],
        @"location" : calendar.location ?: [NSNull null],
        @"organizer" : calendar.organizer ?: [NSNull null],
        @"status" : calendar.status ?: [NSNull null],
        @"summary" : calendar.summary ?: [NSNull null],
        @"start" : @(calendar.start.timeIntervalSince1970),
        @"end" : @(calendar.end.timeIntervalSince1970)
    };
}

@end

File: packages/google_mlkit_barcode_scanning/ios/google_mlkit_barcode_scanning/Sources/google_mlkit_barcode_scanning/GoogleMlKitBarcodeScanningPlugin.swift
import Flutter
import MLKitVision
import MLKitBarcodeScanning
import google_mlkit_commons

@objc
public class GoogleMlKitBarcodeScanningPlugin: NSObject, FlutterPlugin {
  private var instances: [String: BarcodeScanner] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "google_mlkit_barcode_scanning",
      binaryMessenger: registrar.messenger()
    )
    let instance = GoogleMlKitBarcodeScanningPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "vision#startBarcodeScanner":
      handleDetection(call: call, result: result)
    case "vision#closeBarcodeScanner":
      if let args = call.arguments as? [String: Any], let uid = args["id"] as? String {
        instances.removeValue(forKey: uid)
      }
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func initialize(call: FlutterMethodCall) -> BarcodeScanner? {
    guard let args = call.arguments as? [String: Any],
          let array = args["formats"] as? [NSNumber] else { return nil }
    var formats: Int = 0
    for num in array {
      formats += num.intValue
    }
    let options = BarcodeScannerOptions(formats: BarcodeFormat(rawValue: formats))
    return BarcodeScanner.barcodeScanner(options: options)
  }

  private func handleDetection(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let imageData = args["imageData"] as? [String: Any],
          let uid = args["id"] as? String else {
      result(FlutterError(code: "invalid_args", message: "Missing arguments", details: nil))
      return
    }
    guard let image = VisionImage.visionImage(from: imageData) else {
      result(FlutterError(code: "invalid_image", message: "Invalid or missing image data", details: nil))
      return
    }

    let barcodeScanner: BarcodeScanner
    if let existing = instances[uid] {
      barcodeScanner = existing
    } else {
      guard let newScanner = initialize(call: call) else {
        result(FlutterError(code: "invalid_args", message: "Invalid options", details: nil))
        return
      }
      barcodeScanner = newScanner
      instances[uid] = barcodeScanner
    }

    barcodeScanner.process(image) { barcodes, error in
      if let error = error as NSError? {
        result(FlutterError(code: "Error \(error.code)", message: error.domain, details: error.localizedDescription))
        return
      }
      guard let barcodes = barcodes else {
        result([])
        return
      }
      result(barcodes.map { self.barcodeToDictionary($0) })
    }
  }

  private func barcodeToDictionary(_ barcode: Barcode) -> [String: Any] {
    var dictionary: [String: Any] = [
      "type": barcode.valueType.rawValue,
      "format": barcode.format.rawValue,
      "rect": [
        "left": Double(barcode.frame.origin.x),
        "top": Double(barcode.frame.origin.y),
        "right": Double(barcode.frame.origin.x + barcode.frame.size.width),
        "bottom": Double(barcode.frame.origin.y + barcode.frame.size.height)
      ] as [String: Double]
    ]
    dictionary["rawValue"] = barcode.rawValue ?? NSNull()
    dictionary["rawBytes"] = barcode.rawData ?? NSNull()
    dictionary["displayValue"] = barcode.displayValue ?? NSNull()
    let points = (barcode.cornerPoints ?? []).map { point -> [String: Double] in
      let cgPoint = point.cgPointValue
      return ["x": Double(cgPoint.x), "y": Double(cgPoint.y)]
    }
    dictionary["points"] = points

    switch barcode.valueType {
    case .wiFi:
      if let wifi = barcode.wifi { dictionary.merge(wifiToDictionary(wifi)) { _, new in new } }
    case .URL:
      if let url = barcode.url { dictionary.merge(urlToDictionary(url)) { _, new in new } }
    case .email:
      if let email = barcode.email { dictionary.merge(emailToDictionary(email)) { _, new in new } }
    case .phone:
      if let phone = barcode.phone { dictionary.merge(phoneToDictionary(phone)) { _, new in new } }
    case .SMS:
      if let sms = barcode.sms { dictionary.merge(smsToDictionary(sms)) { _, new in new } }
    case .geographicCoordinates:
      if let geo = barcode.geoPoint { dictionary.merge(geoPointToDictionary(geo)) { _, new in new } }
    case .driversLicense:
      if let license = barcode.driverLicense { dictionary.merge(driverLicenseToDictionary(license)) { _, new in new } }
    case .contactInfo:
      if let contact = barcode.contactInfo {
        dictionary.merge(contactInfoToDictionary(contact)) { _, new in new }
      }
    case .calendarEvent:
      if let calendar = barcode.calendarEvent {
        dictionary.merge(calendarEventToDictionary(calendar)) { _, new in new }
      }
    default:
      break
    }
    return dictionary
  }

  private func wifiToDictionary(_ wifi: BarcodeWifi) -> [String: Any] {
    [
      "ssid": wifi.ssid ?? NSNull(),
      "password": wifi.password ?? NSNull(),
      "encryption": wifi.type.rawValue
    ]
  }

  private func urlToDictionary(_ url: BarcodeURLBookmark) -> [String: Any] {
    [
      "title": url.title ?? NSNull(),
      "url": url.url ?? NSNull()
    ]
  }

  private func emailToDictionary(_ email: BarcodeEmail) -> [String: Any] {
    [
      "address": email.address ?? NSNull(),
      "body": email.body ?? NSNull(),
      "subject": email.subject ?? NSNull(),
      "emailType": email.type.rawValue
    ]
  }

  private func phoneToDictionary(_ phone: BarcodePhone) -> [String: Any] {
    [
      "number": phone.number ?? NSNull(),
      "phoneType": phone.type.rawValue
    ]
  }

  private func smsToDictionary(_ sms: BarcodeSMS) -> [String: Any] {
    [
      "number": sms.phoneNumber ?? NSNull(),
      "message": sms.message ?? NSNull()
    ]
  }

  private func geoPointToDictionary(_ geo: BarcodeGeoPoint) -> [String: Any] {
    [
      "longitude": geo.longitude,
      "latitude": geo.latitude
    ]
  }

  private func driverLicenseToDictionary(_ license: BarcodeDriverLicense) -> [String: Any] {
    [
      "firstName": license.firstName ?? NSNull(),
      "middleName": license.middleName ?? NSNull(),
      "lastName": license.lastName ?? NSNull(),
      "gender": license.gender ?? NSNull(),
      "addressCity": license.addressCity ?? NSNull(),
      "addressStreet": license.addressStreet ?? NSNull(),
      "addressState": license.addressState ?? NSNull(),
      "addressZip": license.addressZip ?? NSNull(),
      "birthDate": license.birthDate ?? NSNull(),
      "documentType": license.documentType ?? NSNull(),
      "licenseNumber": license.licenseNumber ?? NSNull(),
      "expiryDate": license.expiryDate ?? NSNull(),
      "issueDate": license.issuingDate ?? NSNull(),
      "country": license.issuingCountry ?? NSNull()
    ]
  }

  private func contactInfoToDictionary(_ contact: BarcodeContactInfo) -> [String: Any] {
    let addresses: [[String: Any]] = (contact.addresses ?? []).map { address in
      [
        "addressLines": address.addressLines ?? NSNull(),
        "addressType": address.type.rawValue
      ] as [String: Any]
    }
    let emails = (contact.emails ?? []).map { (email: BarcodeEmail) -> [String: Any] in
      [
        "address": email.address ?? NSNull(),
        "body": email.body ?? NSNull(),
        "subject": email.subject ?? NSNull(),
        "emailType": email.type.rawValue
      ] as [String: Any]
    }
    let phones = (contact.phones ?? []).map { (phone: BarcodePhone) -> [String: Any] in
      [
        "number": phone.number ?? NSNull(),
        "phoneType": phone.type.rawValue
      ] as [String: Any]
    }
    let name = contact.name
    return [
      "addresses": addresses,
      "emails": emails,
      "phones": phones,
      "urls": contact.urls ?? NSNull(),
      "formattedName": name?.formattedName ?? NSNull(),
      "firstName": name?.first ?? NSNull(),
      "lastName": name?.last ?? NSNull(),
      "middleName": name?.middle ?? NSNull(),
      "prefix": name?.prefix ?? NSNull(),
      "pronunciation": name?.pronunciation ?? NSNull(),
      "suffix": name?.suffix ?? NSNull(),
      "jobTitle": contact.jobTitle ?? NSNull(),
      "organization": contact.organization ?? NSNull()
    ]
  }

  private func calendarEventToDictionary(_ calendar: BarcodeCalendarEvent) -> [String: Any] {
    [
      "description": calendar.eventDescription ?? NSNull(),
      "location": calendar.location ?? NSNull(),
      "organizer": calendar.organizer ?? NSNull(),
      "status": calendar.status ?? NSNull(),
      "summary": calendar.summary ?? NSNull(),
      "start": calendar.start.map { NSNumber(value: $0.timeIntervalSince1970) } ?? NSNull(),
      "end": calendar.end.map { NSNumber(value: $0.timeIntervalSince1970) } ?? NSNull()
    ]
  }
}
